File: Projects/FlutterPerfLab/lib/models/product_data.dart
import 'package:json_annotation/json_annotation.dart';

// Required directive for build_runner to generate compile-time parsing logic.
part 'product_data.g.dart';

/// A raw data schema representing a product entity, optimized for reflection-free JSON deserialization.
///
/// Decoupling the raw decoded properties from view-specific identifiers prevents the
/// generated parser from incurring unnecessary computational overhead during parsing.
@JsonSerializable(createToJson: false)
class ProductData {
  final String name;
  final String price;
  final String oldPrice;
  final String discount;
  final String category;
  final String imageName;

  const ProductData({
    required this.name,
    required this.price,
    required this.oldPrice,
    required this.discount,
    required this.category,
    required this.imageName,
  });

  /// Factory constructor delegating to the static, compile-time generated adapter.
  factory ProductData.fromJson(Map<String, dynamic> json) =>
      _$ProductDataFromJson(json);
}

/// A UI-bound presentation wrapper encapsulating the raw product data.
///
/// Utilizing a deterministic integer identifier bypasses the CPU overhead and Garbage
/// Collection pressure associated with UUID generation, ensuring efficient rendering benchmarks.
class Product {
  final int id;
  final ProductData data;

  const Product({required this.id, required this.data});
}

File: Projects/SwiftPerfLab/Models/ProductData.swift
import Foundation

/// A raw data schema for product entities, designed for efficient JSON deserialization.
/// Decoupling raw decoded properties from view-specific identifiers prevents the JSONDecoder 
/// from incurring unnecessary computational overhead during parsing.
struct ProductData: Decodable {
    let name: String
    let price: String
    let oldPrice: String
    let discount: String
    let category: String
    let imageName: String
}

/// A UI-bound wrapper encapsulating raw product data for rendering.
/// Replaces dynamic `UUID()` allocation with a deterministic integer ID to avoid CPU overhead 
/// during high-frequency scrolling, isolating pure UI rendering performance.
struct Product: Identifiable {
    let id: Int
    let data: ProductData
}